### DISCUSSION TITLE
Response async render

### DISCUSSION THREAD

**Discussion by sector119** [fastapi/fastapi#8463] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class JSONResponse(Response):
    media_type = "application/json"

    def render(self, content: typing.Any) -> bytes:
        return json.dumps(
            content,
            ensure_ascii=False,
            allow_nan=False,
            indent=None,
            separators=(",", ":"),
        ).encode("utf-8")
```

### Description

Hello,

Here is a question. I get JSONResponse as an example, should I and is it possible at all to run render asynchronously?

For example I return some data and want to generate xml response and if this data wouldn't be small it would take some time and block worker? 
Should I run my data2xml function inside render method with starlette.concurrency.run_in_threadpool or something similar to make that call async and not to freeze worker?

And maybe all those json.dumps in render response methods should be called in threadpool too?

Thank You

### Operating System

Linux

### FastAPI Version

0.75.1

### Python Version

3.10.4

**Comment by DeezShiznit** [comment]
> ### First Check
> * [x]  I added a very descriptive title to this issue.
> * [x]  I used the GitHub search to find a similar issue and didn't find it.
> * [x]  I searched the FastAPI documentation, with the integrated search.
> * [x]  I already searched in Google "How to X in FastAPI" and didn't find any information.
> * [x]  I already read and followed all the tutorial in the docs and didn't find an answer.
> * [x]  I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
> * [x]  I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
> * [x]  I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
> 
> ### Commit to Help
> * [x]  I commit to help with one of those options 👆
> 
> ### Example Code
> ```python
> class JSONResponse(Response):
>     media_type = "application/json"
> 
>     def render(self, content: typing.Any) -> bytes:
>         return json.dumps(
>             content,
>             ensure_ascii=False,
>             allow_nan=False,
>             indent=None,
>             separators=(",", ":"),
>         ).encode("utf-8")
> ```
> 
> ### Description
> Hello,
> 
> Here is a question. I get JSONResponse as an example, should I and is it possible at all to run render asynchronously?
> 
> For example I return some data and want to generate xml response and if this data wouldn't be small it would take some time and block worker? Should I run my data2xml function inside render method with starlette.concurrency.run_in_threadpool or something similar to make that call async and not to freeze worker?
> 
> And maybe all those json.dumps in render response methods should be called in threadpool too?
> 
> Thank You
> 
> ### Operating System
> Linux
> 
> ### Operating System Details
> _No response_
> 
> ### FastAPI Version
> 0.75.1
> 
> ### Python Version
> 3.10.4
> 
> ### Additional Context
> _No response_

**Comment by mauroalejandrojm** [comment]
I want to first understand what is the use case of this response async render:
* response from the server to client.
* response from the client to server.

If the response is from the server to the client you can follow the documentation about [websockets](https://fastapi.tiangolo.com/advanced/websockets/#more-info) and change or update the lines inside the loop to continuously receive data:

```
data = await websocket.receive_text()
# Send message to the client
response = {"render":str(json_render.render(content=bytes(data, 'utf-8')))}
await websocket.send_json(response)
```

If the response is from the client to the server you might follow the section about handling disconnections and multiple clients.

If you just want to render the JSON your code might look like this:

```
from fastapi import APIRouter, WebSocket, WebSocketDisconnect
from fastapi.responses import HTMLResponse
from typing import List
import random
import typing
import json

class JSONResponse:
    
    media_type = "application/json"
    
    def __init__(self) -> None:
        pass

    def render(self, content: typing.Any) -> bytes:
        return json.dumps(
            content.decode(),
            ensure_ascii=False,
            allow_nan=False,
            indent=None,
            separators=(",", ":"),
        ).encode("utf-8")

class ConnectionManager:
    def __init__(self):
        self.active_connections: List[WebSocket] = []

    async def connect(self, websocket: WebSocket):
        await websocket.accept()
        self.active_connections.append(websocket)

    def disconnect(self, websocket: WebSocket):
        self.active_connections.remove(websocket)

    async def send_personal_message(self, message: str, websocket: WebSocket):
        await websocket.send_text(message)

    async def broadcast(self, message: str):
        for connection in self.active_connections:
            await connection.send_text(message)

analytics = APIRouter()
json_render = JSONResponse()
manager = ConnectionManager()

html = """
<!DOCTYPE html>
<html>
    <head>
        <title>Render Test</title>
    </head>
    <body>
        <h1>WebSocket Render Page</h1>
        <h2>Socket ID: <span id="ws-id"></span></h2>
        <div id='messages'>
        </div>
        <script>
            let stateObj = { id: "100" };
            var client_id = Date.now()
            window.history.replaceState(stateObj, "client_id", client_id);
            document.querySelector("#ws-id").textContent = client_id;
            var ws = new WebSocket(`ws://localhost:8000/ws/${client_id}`);
            ws.onmessage = function(event) {
                var messages = document.getElementById('messages')
                var message = document.createElement('div')
                var content = document.createTextNode(event.data)
                message.appendChild(content)
                messages.appendChild(message)
                console.log(content)
            };
        </script>
    </body>
</html>
"""




@analytics.get("/")
async def get():
    return HTMLResponse(html)

@analytics.websocket("/ws/{client_id}")
async def websocket_endpoint(websocket: WebSocket, client_id: int):
    await manager.connect(websocket)
    try:
        while True:
        # data = await websocket.receive_text()
        # await websocket.send_text(f"Message text was: {data}")
            data = await websocket.receive_text()
            # Send message to the client
            # resp = {'value': random.uniform(0, 1)}
            # await websocket.send_json(resp)
            response = {"render":str(json_render.render(content=bytes(data, 'utf-8')))}
            await manager.send_personal_message(str(response), websocket)
            await manager.broadcast(f"Client #{client_id} sent: {str(data)}")
    except WebSocketDisconnect:
        manager.disconnect(websocket)
        await manager.broadcast(f"Client #{client_id} left")

```

and the client should be rendering something like this, Note that Tiagolo declares a client_id based on the `Date.now()` method of Javascript, but you can set it to a fixed client_id:

```
<body data-new-gr-c-s-check-loaded="14.1011.0" data-gr-ext-installed="">
        <h1>WebSocket Render Page</h1>
        <h2>Socket ID: <span id="ws-id">1649927783794</span></h2>
        <div id="messages">
        <div>Client #1649927783794 sent: testing render</div></div>
        <script>
            let stateObj = { id: "100" };
            var client_id = Date.now()
            window.history.replaceState(stateObj, "client_id", client_id);
            document.querySelector("#ws-id").textContent = client_id;
            var ws = new WebSocket(`ws://localhost:8000/ws/${client_id}`);
            ws.onmessage = function(event) {
                var messages = document.getElementById('messages')
                var message = document.createElement('div')
                var content = document.createTextNode(event.data)
                message.appendChild(content)
                messages.appendChild(message)
                console.log(content)
            };
        </script>
    

</body>
```

Notice I changed the `render` function in the `JSONResponse` to serialize the data and render it in the HTML.
Regarding the other questions about bigger data, I don't think they will block the broker, I tested a random generator and render all the data and the HTML was rendering ok.

**Comment by chbndrhnns** [comment]
When you create the response, all the data required to do so should already be available. Retrieving the data could happen inside the endpoint method. When all data is present, there is no blocking I/O to do and thus no need to think further about the method in `JSONResponse`.

**Comment by sector119** [comment]
JSONResponse.render have no blocking IO, but, for example I want to have some XMLRPC response from server to client and I return some really big list of structures and conversation of this object to xml take some time, would it block loop? Wouldn’t it be better in that case to run that conversation in threadpool to skip loop blocking?

**Comment by chbndrhnns** [comment]
That depends on the mechanism you use to retrieve that data. If the mechanism is using async, there is no need to use threads. Otherwise, yes. But I still think this would happen outside the response class.

**Comment by sector119** [comment]
I decided to write my own XMLRPCResponse class based on starlette Response and just override render method to put there myxmlrpclib.dump(content) where content is data returned by endpoint like JSONResponse does.

I don’t know if it is possible to make myxmlrpclib.dump(content) async because it is mostly cpu bound operation.

Or maybe I can stream xml during conversation from Response.__call__ with multiple send calls?

**Comment by YuriiMotov** [comment, marked as answer]
`Response.render` method is called by `Response.__init__`, that is `sync`.

If you are concerned that your response being too large and that it may block the worker, consider using `StreamingResponse` (https://fastapi.tiangolo.com/advanced/custom-response/#streamingresponse)
